### ISSUE TITLE
sql: tuples with NULLs don't compare sanely

### ISSUE THREAD

**Issue by knz** [cockroachdb/cockroach#12022]
Found by @eisenstatdavid in https://github.com/cockroachdb/cockroach/pull/10475#issuecomment-264911948

> The logic for not defining the previous element of (e.g.) -inf to be null is that, in a filter expression like col < -inf, the null values of col don't satisfy the filter (because null < -inf is null (unknown, technically, but we conflate unknown and boolean null) when we're using the user-accessible builtin and not .Compare), and thus should be excluded.
>
> For arrays, I think Next(a) = a + [null] is good enough for the limited purpose of scanning an index. If we unthinkingly rewrite y > [1] to y >= [1, null], then despite the fact that [1, 2] > [1] is true, [1, 2] > [1, null] should be unknown (guessing this is what PostgreSQL does from its handling of tuples; IIRC, we don't do this because the array builtin calls .Compare, but we should).
>
> For tuples also, we don't have a good answer. (2, null) sorts in between (1, max) and (2, min), so we should include (2, null) when there's a > (1, max) filter. This means that defining Next((1, max)) = (2, min) is bad, because then it's not safe to rewrite, e.g., y > (1, max) to y >= (2, min), because the semantics for (2, null) changes, from true to either false (.Compare) or null (as the builtin should return, but doesn't yet). Defining Next((1, max)) = (2, null) has the problem described above.

Jira issue: CRDB-6132

**Comment by nvanbenschoten** [comment]
I'm not sure that we ever want to return NULL values for a query that has a predicate on those values. The logic for this is that the comparison would be unknown, like you pointed out, so the null value would never pass the filter. This is how our current ordering works (with and without indexes, just like PG), and I believe means that this issue is not needed.

```
root@:26257> select a, b from t order by b;
+---+-------+
| a |   b   |
+---+-------+
| 3 | NULL  |
| 1 | false |
| 2 | true  |
+---+-------+
(3 rows)
root@:26257> select a, b from t where b > false order by b;
+---+------+
| a |  b   |
+---+------+
| 2 | true |
+---+------+
(1 row)
root@:26257> select a, b from t where b <= false order by b;
+---+-------+
| a |   b   |
+---+-------+
| 1 | false |
+---+-------+
(1 row)
```

Following this logic, the new issue would be making sure that that `(3, null)` is not returned from a query with a predicate of `y > (1, max)`.

**Comment by eisenstatdavid** [comment]
Tuple comparison is lexicographic, not pointwise, hence PostgreSQL does return true sometimes when comparing tuples with null:

```
eisen=# SELECT (1, 'inf'::float) < (3, null);
 ?column?
----------
 t
(1 row)
```

**Comment by nvanbenschoten** [comment]
Does that change anything with respect to the initial question?

```
nathan=# SELECT (2, 'inf'::float) < (2, null);
 ?column?
----------

(1 row)
```

**Comment by eisenstatdavid** [comment]
That's the reason we can't rewrite y > (1, max) to y >= (2, null).

**Comment by eisenstatdavid** [comment]
> Interestingly, normalizing y > (1, max) to y >= (1, null) would work because of what I mentioned earlier.

Only if we don't use an index, which puts null first. We could go deeper and rewrite to y' >= (2,), dropping the last component out.

**Comment by nvanbenschoten** [comment]
I might be missing something, but it doesn't look like we ever use indexes for inequalities between tuples. @RaduBerinde do you know if this is true, and if so, why we don't currently use them?

```
root@:26257> EXPLAIN SELECT a, b FROM t WHERE (a, b) = (2, true);
+-------+------+-------------------------+
| Level | Type |       Description       |
+-------+------+-------------------------+
|     0 | scan | tt@tt_a_b_idx /2/1-/2/2 |
+-------+------+-------------------------+
(1 row)
root@:26257> EXPLAIN SELECT a, b FROM t WHERE (a, b) > (2, true);
+-------+------+-----------------+
| Level | Type |   Description   |
+-------+------+-----------------+
|     0 | scan | tt@tt_a_b_idx - |
+-------+------+-----------------+
(1 row)
```

Regardless, dropping out the last component during index selection like you brought up seems like the correct approach. This is what I was hoping to achieve with the hacky normalization from y > (1, max) to y >= (1, null). 

We'll also need to take note that index selection currently adds in an implicit IS NOT NULL constraint for isolated end constraints, resulting in the behavior I described above. If/when we support the use of indexes for inequalities between tuples, similar behavior will be expected.

**Comment by RaduBerinde** [comment]
There are various known gaps in the span generation code (we have some issues filed #6390, #6346).

**Comment by eisenstatdavid** [comment]
Concrete test case:
```
CREATE DATABASE d;
SET DATABASE = d;

CREATE TABLE t (
  c1 INT,
  c2 INT,
  UNIQUE INDEX i (c1, c2)
);

INSERT INTO t VALUES
  (NULL, NULL), (NULL, 1), (NULL, 2),
  (1, NULL), (1, 1), (1, 2),
  (2, NULL), (2, 1), (2, 2);

SELECT * FROM t WHERE (c1, c2) > (1, 9223372036854775807);
```
returns
```
c1	c2
2	1
2	2
```
but `2 NULL` should be included.

**Comment by eisenstatdavid** [comment]
The PR references above implements Nathan's "hacky normalization" strategy, documenting at length in the contract for Next/Prev why it works for now.

**Comment by knz** [comment]
@awoods187 I am bumping this issue because it's actually a correctness issue that's likely to bite us earlier than later. I am not sure how to best approach it though wrt a solution. This might need input from multiple people (Nathan, Peter, Ben, Radu come to mind).

**Comment by awoods187** [comment]
@knz i'm on board with prioritizing correctness. Do we need a full rfc here?

**Comment by knz** [comment]
I don't know actually. I do not understand the scope of the problem (i.e. how deep we're screwed).

**Comment by knz** [comment]
In doubt, I'd say yes we need a RFC.

**Comment by awoods187** [comment]
@vivekmenezes let's talk about this later today

**Comment by lego** [comment]
This is not completely related, but to add one another data point of an inconsistency with NULL usage is our usage of `Datum.Compare` with two `DNull`. This evaluates to true and if misused in a place where we do not know if the datums are non-NULL we will end up getting true whereas the SQL evaluation would be false-y.

**Comment by RaduBerinde** [comment]
`Datum.Compare` should work like that, otherwise we wouldn't be able to sort them correctly (we advertise "NULLs first" in orderings). `ComparisonExpr`, which is what gets evaluated in an expression, does check for NULLs. https://github.com/cockroachdb/cockroach/blob/99fe767f728d85d6fd1e7a8ba715a8e5e9b6860f/pkg/sql/sem/tree/eval.go#L1917

**Comment by lego** [comment]
I meant for that to be more of a commentary on the internal usage of `Datum.Compare`. I don't know if anything can be done to help prevent mistakes. (And I brought it to this issue since this seemed partially related)

**Comment by arjunravinarayan** [comment]
This is obviously not getting done for 2.0. Bumping milestone to 2.1, @awoods187 pleased triage and ensure there is an Airtable entry for this, or remove the 2.1 milestone as well.

**Comment by awoods187** [comment]
We have it slotted in airtable for 2.1. Good catch on bumping this here.

**Comment by arjunravinarayan** [comment]
Thank you for confirming @awoods187!

**Comment by knz** [comment]
@vivekmenezes I am ok with looking at this, but this item requires a team of two or more engineers.

Also 3/4 of the work will be on the SQL execution side, not SQL language.

I would seriously prefer Nathan as work buddy on this, but anyone with deeper understanding of all aspects of SQL execution would be fine as well. Please also consider invoking Radu to provide guidance on the work, as it moves along.

**Comment by nvanbenschoten** [comment]
@knz I love the idea of working as a team on this! I'd be happy to work together during one of the weeks that you're in the office in the 2.1 cycle to remedy this.

**Comment by knz** [comment]
The damage on this issue is currently somewhat limited.

For arrays, we're fine because:

- inequality comparison operators are not defined on array types
- it's not possible to ORDER BY an array type
- it's not possible to create an index on array types.

So arrays are "safe".

For tuples, the damage is a little contained because it is not possible to store a value of tuple type, and not possible to define an index over it.

However, I am readily able to reproduce inconsistent results as initially reported in this issue, as follows:

```
> select * from (values(row(0,0)),(row(1,0)),(row(1,null::int)),(row(1,2))) as a(v) order by v;
+-------+
|   v   |
+-------+
| (0,0) |
| (1,)  |
| (1,0) |
| (1,2) |
+-------+

> select * from (values(row(0,0)),(row(1,0)),(row(1,null::int)),(row(1,2))) as a(v) where v < row(1,0);
+-------+
|   v   |
+-------+
| (0,0) |
+-------+
``` 

Notice how the first result (using ORDER BY and thus Compare) properly sorts the NULLs, whereas the second one uses the built-in function which returns the wrong result.

**Comment by knz** [comment]
@RaduBerinde I am especially concerned about this issue because I fear that query decorrelation of things like `where (1,1) < (select x,y from foo)` will return silently bogus results if there are nulls in column y.

**Comment by knz** [comment]
Another inconsistency:

```
> select (1,0) < (1,null);
+----------+
| ?column? |
+----------+
|   NULL   |
+----------+

> select (1,(0,0)) < (1,(0,null));
+----------+
| ?column? |
+----------+
|  false   |
+----------+
```

I have no idea where this comes from.

**Comment by knz** [comment]
I am fixing the most egregious problems here in #27885. However there's something strange remaining, look what I found using postgres:

```
kena=# select row(1::int,0::int) < row(1::int,null::int);
 ?column?
----------
 NULL

kena=# select row(1::int,0::int) > row(1::int,null::int);
 ?column?
----------
 NULL
(1 row)
```

(so far, unsurprising)

Now check this:

```
kena=# select v < row(1::int,null::int) from (values(row(1::int,0::int))) as t(v);
 ?column?
----------
 t

kena=# select v > row(1::int,null::int) from (values(row(1::int,0::int))) as t(v);
 ?column?
----------
 f
(1 row)
```

So one would now think the result of the inequality comparison is different whether the operand is a tuple expression or a column variable. This finding would already be troublesome, but that's not all there is to it!

The behavior is special even when column variables are "embedded" inside a more complex expression. Like this:

```
kena=# select row(0, row(1, 0)) < row(0, x) from (values(row(1, null::int))) as t(x);
 ?column?
----------
 t
(1 row)
```

My conclusion from this observation is that **column vars in scalar comparisons are always totally ordered with other values, regardless of whether their value for the current row is NULL*

And it's not only inequality, even equality has this:

```
kena=# select row(0, row(1, 0)) = row(0, x) from (values(row(1, null::int))) as t(x);
 ?column?
----------
 f


kena=# select row(0, row(1, null::int)) = row(0, x) from (values(row(1, null::int))) as t(x);
 ?column?
----------
 t
```

so we can "find" NULLs with SQL equality in tuples if the NULLs are "hidden" behind a column variable.

cc @RaduBerinde @nvanbenschoten

**Comment by knz** [comment]
This in turn explains why `select ... where x > y` in my comment above https://github.com/cockroachdb/cockroach/issues/12022#issuecomment-405221490 is well-defined in postgres: the comparison uses total order even though some of the values contain nulls.

**Comment by RaduBerinde** [comment]
> The behavior is special even when column variables are "embedded" inside a more complex expression. Like this:

```
kena=# select row(0, row(1, 0)) < row(0, x) from (values(row(1, null::int))) as t(x);
 ?column?
----------
 t
(1 row)
```
I don't think this case has anything to do with column variables; it looks like this is always the behavior with deeper tuples:
```
radu=> select row(1, row(1, 0)) < row(1, row(1, NULL::int));
 ?column? 
----------
 t
(1 row)
```

I wonder if some of these cases may be bugs rather than intended behavior.

**Comment by knz** [comment]
Ok let's ignore the "values inside tuples" case. What do you suggest to do about this:

```
kena=# select v < row(1::int,null::int) from (values(row(1::int,0::int))) as t(v);
 ?column?
----------
 t
```

Which makes it possible in pg to have:

```
kena=# select * from (values(row(0,0)),(row(1,0)),(row(1,null::int)),(row(1,2))) as a(v) where v > row(1,0);
   v
-------
 (1,)
 (1,2)
```

(remember, in pg NULLs sort last)

whereas in CockroachDB currently (where NULLs sort first):

```
b> select * from (values(row(0,0)),(row(1,0)),(row(1,null::int)),(row(1,2))) as a(v) where v < row(1,0);
+-------+
|   v   |
+-------+
| (0,0) |
+-------+
```

(we'd want `(1, NULL)` in the results here)

**Comment by knz** [comment]
Note that this result would be different in CockroachDB if we use distsql, use key encoding for the tuple and use comparisons on the key encoding for the boundary.

**Comment by knz** [comment]
Update: Radu and I talked and we've agreed that the "Special" behavior in postgres is likely an oversight and according to the SQL standard is likely to be a bug.

In any case pg does not document it! And its docs instead strongly suggest that it should behave differently than it does.

So we'll implement the "regular" behavior (NULL comparisons yield NULL) which will only differ from postgres when there are sub-tuples or nulls inside arrays.

We may want to document this but then again pg doesn't either.

**Comment by knz** [comment]
pg bug/question filed upstream as bug 15297

https://www.postgresql.org/message-id/153261212701.1395.18260320763646237974%40wrigleys.postgresql.org

**Comment by knz** [comment]
pg bug/question specific to arrays filed upstream as bug 15298

https://www.postgresql.org/message-id/153261258644.1404.3005762345361971551%40wrigleys.postgresql.org

**Comment by knz** [comment]
The conversation on the upstream issue is instructive

https://www.postgresql.org/message-id/153261212701.1395.18260320763646237974%40wrigleys.postgresql.org

Basically the argument is "we don't do SQL standard NULLs because otherwise we can't use Btrees and it's not fast"  and "for top-level tuples oh well we have a weird behavior but that's just historical so now we can't change it any more"

... which kinda holds together.

In short, pg decided to violate the sql standard for performance, and is stuck not being able to do this for top-level tuples/arrays because of backward compat.

**Comment by knz** [comment]
I'll quote Tom Lane from the pg list:

> if we have a comparison operator comparing two **syntactic** [emphasis mine] row
> constructors, it works differently: that case breaks down the two field
> lists and applies the named operator to each pair of values.

So if the *syntax* says `(a, b) = (c, d)` then we have the NULL-sensitive special behavior, whereas if the *syntax* says `x = (c, d)` and `x` has composite type, we don't have the special behavior and we use total order.

**Comment by knz** [comment]
@RaduBerinde I want to suggest creating a built-in AST node TupleCompareExpr for tuple-tuple comparisons, so that any expression of the form `(..x..) CMP (..y..)` get handled with that AST node instead of the regular ComparisonExpr.

This is because every *other* use of ComparisonExpr will use the "fast" (non-standard) behavior and I prefer we don't have special code at every point that looks at ComparisonExpr just for tuples.

What do you think?

**Comment by knz** [comment]
This is especially important because after semantic analysis, `CompareExpr` cannot distinguish between a direct sub-operand that was a *syntactic* tuple from one that just happens to be a tuple (IndexedVar from a tuple-typed column)

**Comment by knz** [comment]
... Also pretty-printing for distsql proc specs will not emit the right thing in some cases unless we use a different syntax.

That also suggests we need some kind of explicit syntax for this new node (Again for distsql proc specs)

**Comment by knz** [comment]
Discussed with @RaduBerinde - a separate AST node has several benefits in addition to those already listed in the comment already. I'll try something like that.

**Comment by justinj** [comment]
A difference from Postgres:

Postgres:
```
select (1, null) is null, (1, null) is not null, (1, null) is distinct from null, (1, null) is not distinct from null;
 ?column? | ?column? | ?column? | ?column?
----------+----------+----------+----------
 f        | f        | t        | f
(1 row)
```

Cockroach:
```
> select (1, null) is null, (1, null) is not null, (1, null) is distinct from null, (1, null) is not distinct from null;
  ?column? | ?column? | ?column? | ?column?
+----------+----------+----------+----------+
  NULL     | NULL     | NULL     | NULL
(1 row)
```

**Comment by RaduBerinde** [comment]
Hm, that looks like a bug. `IS NULL` and similar should only return true or false.

**Comment by jordanlewis** [comment]
@justinj's issue no longer reproduces, nor does David Eisenstat's original reproduction.

@knz, if you still have context, could you please provide a minimal reproduction of the problem here? I will edit the main body to include that reproduction. If there is no longer a reproduction, let's close this issue and re-open with something more concrete.

**Comment by knz** [comment]
This PR was meant to fix it: https://github.com/cockroachdb/cockroach/pull/27885

I was not able to complete the PR back then because of limitations in the type system. I think the work can be resumed now that the code has been greatly simplified by Andy.

At any rate the examples/tests in the PR give you an *idea* of what's problematic. Note that the examples/tests in the PR also happen to be currently *incorrect* -- they are "reasonable" (I wrote them by extending the original problem scenarios in a way that was consistent and symmetric) but then Radu and I then later discovered that postgres actually diverges from what's reasonable. I even posted on the pg mailing list to ask "wtf" and the answer was "historical reasons". 

To summarize:
- current crdb behavior both inconsistent and different from pg
- PR #27885 makes behavior consistent and regular
- pg behavior actually different from #27885 (albeit less inconsistent than current crdb behavior).

I'm not paged in on this today but we can discuss when I'm done with this week's other concurrent activities.

**Comment by jordanlewis** [comment]
Having reviewed this issue again, I'm going to rename it to be specific to tuple comparisons, which seem to still be somewhat messed up in the presence of nulls. @knz, if you disagree with this or have further context I'm missing, I'll politely request that you open a separate issue that's scoped just to the thing that's broken here besides tuple comparison (which I still can't figure out).

Postgres:
```
jordan=# select (1, null) is null, (1, null) is not null, (1, null) is distinct from null, (1, null) is not distinct from null;
 ?column? | ?column? | ?column? | ?column?
----------+----------+----------+----------
 f        | f        | t        | f
(1 row)
```

Cockroach:
```
root@127.0.0.1:54289/d> select (1, null) is null, (1, null) is not null, (1, null) is distinct from null, (1, null) is not distinct from null;
  ?column? | ?column? | ?column? | ?column?
+----------+----------+----------+----------+
   false   |   true   |   true   |  false
(1 row)
```

**Comment by jordanlewis** [comment]
@yuzefovich this is related to recent work you were doing I think. Can you merge into the other one if appropriate?

**Comment by yuzefovich** [comment]
I handed over #46675 to the optimizer team. Regarding merging this issue with that one, I'm worried that this issue is a superset of #46675, so I don't want to close this.

**Comment by jordanlewis** [comment]
None of the examples in this issue repro anymore, so I'm going to close this.
